Make WooCommerceScreenshots work again on iPhone tests

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Closes: #https://github.com/woocommerce/woomobile-private/issues/392
<!-- Id number of the GitHub issue this PR addresses. -->

`testScreenshots()` in `WooCommerceScreenshots` will fail when being run because since the last time it is updated, many more actions were added that have not been mocked yet.

This PR aims to fix that by adding the proper mocks and going through the test steps and updating what's missing. There are quite a bit of commits, but most of them are individual mocking of a single action to make it clearer what changes are needed.

For comparison's sake, something like this has been done before in previous PRs https://github.com/woocommerce/woocommerce-ios/pull/5123 https://github.com/woocommerce/woocommerce-ios/pull/6699 . However, this PR is much bigger because since it's been years since the update is done, and there are many more mocks to be added.

**Important note**: Please note while the screenshot taking is done as a test `XCTestCase`, the final purpose of it is to create five exact screenshots as shown in the App Store below, **not to test realistic usage flow in the app**:

![Screenshot 2024-11-12 at 22 22 12](https://github.com/user-attachments/assets/f69308e4-8cdc-4b94-bd16-fcd6dd62783f)

![Screenshot 2024-11-12 at 22 22 19](https://github.com/user-attachments/assets/48caead1-01cc-4a23-a33c-de577e3b11e7)

To arrive at each of those screens,`WooCommerceScreenshots` try to "cheat" and do it as quickly as possible, by:
- Mocking and ignoring things that are not essential to the needed screens,
- Skipping various steps that would have been needed in actual use.

Thus, this PR should mostly be decided on whether it's able to generate the correct screenshots, and less about how it does so in the code.

Also, checking and handling for iPad cases will be done separately.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Select `WooCommerceScreenshots` as target,
2. Go to Test Navigator, search for `testScreenshots()`
3. Run `testScreenshots()`, ensure that this opens the Simulator and starts going through screens automatically, and it finishes completely with the test passing.
4. Check `[homedir]/Library/Caches/tools.fastlane/screenshots` and ensure there are generated screenshots there. You might need to create this directory first if you don't have it.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested these using Xcode 16.1, with two different simulators:
1. iPhone 15 with iOS 18
2. iPhone 8 with iOS 16

Fastlane uses the later, while iPhone 15 with iOS 18 is the latest available Simulator in Xcode 16.1.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

Please note that the screenshots below will differ in some elements compared to the App Store screenshots above. There has been various design changes since the last time those screenshots were created.

**on iPhone 15 iOS 18:**

| Dashboard | Order Creation | Order Payment | Product Creation | Order Notification |
|-|-|-|-|-|
| ![Hafiz iPhone 15-1-light-order-dashboard](https://github.com/user-attachments/assets/f5352c7b-40fd-47d6-abd8-0d97d05533d3) | ![Hafiz iPhone 15-2-light-order-creation](https://github.com/user-attachments/assets/2dc6c8d1-a124-42dd-91b3-21a0c8f06315) | ![Hafiz iPhone 15-3-light-order-payment](https://github.com/user-attachments/assets/14fdb29a-efd7-47e6-9303-c0c20a53ec84)  | ![Hafiz iPhone 15-4-light-product-add](https://github.com/user-attachments/assets/18d648cf-b816-4e1a-b7d4-cdfd6f41c670) | ![Hafiz iPhone 15-5-light-order-notification](https://github.com/user-attachments/assets/17a62f23-062d-4be5-aa29-05854d2a5792) |

**on iPhone 8 iOS 16**
| Dashboard | Order Creation | Order Payment | Product Creation | Order Notification |
|-|-|-|-|-|
| <img width="375" alt="Hafiz iPhone 8 iOS 16 0-1-light-order-dashboard" src="https://github.com/user-attachments/assets/14840442-decd-4f6a-bc63-526d05ae4ef4"> | <img width="375" alt="Hafiz iPhone 8 iOS 16 0-2-light-order-creation" src="https://github.com/user-attachments/assets/08149962-cc78-4e22-8bc0-20a7b6017270"> | <img width="375" alt="Hafiz iPhone 8 iOS 16 0-3-light-order-payment" src="https://github.com/user-attachments/assets/e481d5a3-63be-475f-ac10-aa86f3cbc619"> | <img width="375" alt="Hafiz iPhone 8 iOS 16 0-4-light-product-add" src="https://github.com/user-attachments/assets/e3809f49-4b8f-41a3-9914-60e9a12962cf"> | <img width="375" alt="Hafiz iPhone 8 iOS 16 0-5-light-order-notification" src="https://github.com/user-attachments/assets/1352e75b-0c7d-42d3-9dff-c7d0429bb948"> |

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hafizrahman committed:
First pass at removing fatal error about non-mocked actions in WooCommerceScreenshots.

The newly added mocked actions do not yet do anything, and need more updates to get the app to show the right contents.

hafizrahman committed:
Add proper mock for `getSkippedCashOnDeliveryOnboardingStep`

hafizrahman committed:
Add proper mock for `loadSiteHasAtLeastOneIPPTransactionFinished`

hafizrahman committed:
Add proper mock for `loadLastSelectedPerformanceTimeRange`

hafizrahman committed:
Add proper mock for `loadLastSelectedTopPerformersTimeRange`

hafizrahman committed:
Add proper mock for `loadLastSelectedMostActiveCouponsTimeRange`

hafizrahman committed:
Add proper mock for `loadLastSelectedStockType`

hafizrahman committed:
Add proper mock for `loadLastSelectedOrderStatus`

hafizrahman committed:
Add proper mock for `loadFavoriteProductIDs`

hafizrahman committed:
Add proper mock for `loadCardReader`

hafizrahman committed:
First pass at displaying Dashboard cards Last Orders and Reviews.

Cards chosen because they were the easiest to mock.

hafizrahman committed:
First pass at making WooCommerceScreenshots not crash due to un-mocked action on Orders screen.

hafizrahman committed:
First pass at making the entire testScreenshots() complete successfully.

This is done by commenting out parts that still need fixing, marked with TODOs. They will be fixed next.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/232">21.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Remove productsNavigationBarGetter as it causes test to fail.

hafizrahman committed:
Add proper mock for `syncSite`

hafizrahman committed:
First pass at displaying Performance card on screenshot test.

hafizrahman committed:
Fix StatsStoreV4 deallocation in MockStatsActionV4Handler

When the StatsStoreV4 instance was being created as a local variable in retrieveStats(), etc, it caused it  to be deallocated before the async CoreData  operation could complete. This resulted in stats data not being saved
due to [weak self] returning early in `StatsStoreV4.upsertStoredOrderStats`.

This commit makes StatsStoreV4 a property of MockStatsActionV4Handler to ensure it remains alive throughout async operations. In effect, this allows the Performance card in the screenshot test to show correct order data.

hafizrahman committed:
Update screenshot to snap the Dashboard right away since "This Month" tab does not exist anymore (replaced by the various stats cards).

"order-dashboard" is no longer accurate as a name but don't want to rename it in case Fastlane uses it when generating App Store screenshots.

hafizrahman committed:
First pass at making Card Present Payment screenshot flow working.

This now makes the payment options appear, but need more handling for a new banner screen "Do you want to add Pay In Person to your web checkout"

hafizrahman committed:
Try: make it possible to bypass cash on delivery setup modal that appears at the beginning.

hafizrahman committed:
Add mocking for .observeConenctedReaders and .retrieveOrderRemotely to advance the IPP modal state.

hafizrahman committed:
Add proper mock for .collectPayment() to immediately show the CardPresentModalTapCard as needed by the screenshot.

hafizrahman committed:
Update oncardReaderMessage mock to be more accurate and update the flow in WooCommerceScreenshots

hafizrahman committed:
Remove unneeded skipping parameter now that the card mocking is done differently.

hafizrahman committed:
Fix lockScreen notification display in WooCommerceScreenshots

hafizrahman committed:
Merge branch 'trunk' into task/14325-woocommercescreenshots

hafizrahman committed:
Force showing no AI product creation option for screenshot purposes.

hafizrahman committed:
Fix checking for -disable-product-creation-ai argument.

hafizrahman committed:
Update logic to ensure Performance card in screenshot mode shows visit and conversion data correctly.

hafizrahman committed:
Consolidate screenshot launch argument to re-use `screenshot-mode` for hiding AI option.

hafizrahman committed:
Add optional lockScreen() handling depending on the OS version.

hafizrahman committed:
Some fixes to make the automated flow work on smaller screen (e.g: iPhone 8).

I tested this to still also work on bigger screen (e.g: iPhone 15)

hafizrahman committed:
Update SnapshotHelper

(This was updated automatically when running `bundle exec fastlane screenshots`, which then included a step to update it)

hafizrahman committed:
Restore System Dialog handling as it seems to work fine now after the other mock updates.

hafizrahman committed:
Update various functions to work with the recent updates in SnapshotHelper

hafizrahman committed:
Update the dialog tapper to not search button by text, but by position.

Searching by text fails when the device is in a non-English language and the text is translated.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/StoreStats/StorePerformanceViewModel.swift:
```diff
@@ -160,8 +160,9 @@ final class StorePerformanceViewModel: ObservableObject {
 
-        // Stop if data is relatively new
-        if !forceRefresh && DashboardTimestampStore.isTimestampFresh(for: .performance, at: timeRange.timestampRange) {
+        // Stop if data is relatively new (unless forceRefresh or in screenshot mode)
+        if !forceRefresh && !ProcessConfiguration.shouldRespectScreenshotMode &&
+           DashboardTimestampStore.isTimestampFresh(for: .performance, at: timeRange.timestampRange) {
```
Could you please explain why we do this in screenshot mode?

hafizrahman replied:
So the issue encountered here is that since screenshots are run multiple times at a close time, subsequent runs will make the `isTimestampFresh()` return true and skip the mock syncing, making the card return empty values:
![Screenshot 2024-11-15 at 16 20 00](https://github.com/user-attachments/assets/cddb4dfe-3b49-4a8e-81cd-3fac882d68ca)

The `shouldRespectScreenshotMode` was a hack to make it do mock syncing all the time. 

The code before this part actually tries to get mocked content, but somehow it's not returning anything.

```
        // Preemptively show any cached content
        periodViewModel?.loadCachedContent()
  ```

I'll try to remove this hack and figure out why the cached content are not found during multiple runs.

hafizrahman replied:
Found the bug, I'll make a separate PR for it as it also affects the app.

hafizrahman replied:
issue: https://github.com/woocommerce/woocommerce-ios/issues/14436

hafizrahman replied:
PR to fix: https://github.com/woocommerce/woocommerce-ios/pull/14438

hafizrahman replied:
This is now solved in https://github.com/woocommerce/woocommerce-ios/pull/14438 and the fix brought in to this PR via https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/ae89fcb05fe293a6b7f7c2032bf5b737a054cef9

hafizrahman replied:
screenshot mode flag is removed in https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/c64a1d3111c045e85a9b4f64268f473b8cf9f192

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/ProductCreationAIEligibilityChecker.swift:
```diff
@@ -20,6 +20,12 @@ final class ProductCreationAIEligibilityChecker: ProductCreationAIEligibilityChe
         }
 
-        return site.isWordPressComStore || site.isAIAssistantFeatureActive
+        if ProcessConfiguration.shouldRespectScreenshotMode {
+            // Do not show AI product creation option in Screenshot Mode.
+            return false
+        }
```
If the new AI option is breaking the navigation, I think we should instead update the UI test flow by adding accessibility identifiers and simulating tap.

rachelmcr replied:
Alternately, we could update the mocks to make the store reflect one that is not eligible, using the checks already in the `ProductCreationAIEligibilityChecker` (not a WordPress.com store, AI assistant feature not active).

hafizrahman replied:
Good point. I was trying not to change the mocked site fearing that it might create unintended breakage elsewhere. But in https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/e80c6380074ec4d4d0602195b6b48038e54ddf7a I set `isWordPressComStore` as false and seems like it's not causing issue anywhere else.

selanthiraiyan reviewed (commented):
> Thus, this PR should mostly be decided on whether it's able to generate the correct screenshots, and less about how it does so in the code.


> For comparison's sake, something like this has been done before in previous PRs https://github.com/woocommerce/woocommerce-ios/pull/5123 https://github.com/woocommerce/woocommerce-ios/pull/6699 . 


I am concerned about the new `screenshot-mode` launch argument. 

I see that we change the code flow by adding conditions if the app is in screenshot mode. In the past screenshots related PRs that you mentioned in the PR description, I don't see us disabling features for taking screenshots. 

Instead of changing the code flow, could we try adding accessibility identifiers and adding/updating the tapping/navigation code in UI tests to achieve this? 

What do you think? 

Tagging @jaclync and @rachelmcr to get their opinion on the above.

rachelmcr commented:
> Instead of changing the code flow, could we try adding accessibility identifiers and adding/updating the tapping/navigation code in UI tests to achieve this?

> What do you think?

I agree with generally trying to avoid changing the code flow to introduce conditions for taking screenshots. I'd expect the screenshots to reflect the actual product, so we should be able to navigate it with UI test interactions and not modify what's shown on screen beyond the mocked data.

hafizrahman committed:
Update Screenshot's site mock to not be a WordPressCom store, removing the need for specific flag for product creation screen.

hafizrahman committed:
Remove usage of .screenshot-mode

hafizrahman committed:
Merge branch 'trunk' into task/14325-woocommercescreenshots

hafizrahman committed:
Merge branch 'trunk' into task/14325-woocommercescreenshots

hafizrahman commented:
@selanthiraiyan I've addressed the feedbacks above, so this is good for another review and test.

selanthiraiyan commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockPaymentActionHandler.swift:
```diff
@@ -0,0 +1,38 @@
+    let objectGraph: MockObjectGraph
+    let storageManager: StorageManagerType
+
+    private let paymentStore: PaymentStore
```
Do we need this store? I don't see this being used anywhere.

hafizrahman replied:
Good catch, removed in https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/c41f3d2e6250b58538eee3f23ebdfbec635ed530

selanthiraiyan commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockSettingActionHandler.swift:
```diff
@@ -23,7 +23,8 @@ struct MockSettingActionHandler: MockActionHandler {
             synchronizeGeneralSiteSettings(siteID: siteID, onCompletion: onCompletion)
-
+        case .retrieveTaxBasedOnSetting:
+            break
```
Could we call the completion block with a mock value? 

I am concerned that if we use `break` here we wouldn't know that this action is not implemented, and this will make debugging difficult for future screenshots related work. 

What do you think? 

The same comment applies to other `MockJustInTimeMessageActionHandler`, `MockGoogleAdsActionHandler`, `MockShippingMethodActionHandler` and other places where we have used `break` to avoid getting into the `unimplementedAction` path.

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/9e53017013e930e5b1e53106c4edfe034c16a3a8

selanthiraiyan commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockAppSettingsActionHandler.swift:
```diff
@@ -23,10 +23,32 @@ struct MockAppSettingsActionHandler: MockActionHandler {
                 .loadOrdersSettings,
-                .upsertProductsSettings:
+                .upsertProductsSettings,
+                .loadFirstInPersonPaymentsTransactionDate
```
Same as my other comment about using `break` to avoid calling `unimplementedAction`. Could we avoid using `break` for actions with a completion block?

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/9e53017013e930e5b1e53106c4edfe034c16a3a8

selanthiraiyan reviewed (commented):
Thanks for working on this, @hafizrahman! 🙇 

This looks good to me in general. 

I am worried about using `break` for actions with completion handlers. If we don't implement those actions by calling the completion handler, it will be difficult to debug it in future. 

What do you think? Will calling the completion blocks (with mock value or even error if needed) change the navigation flows and break the screenshots flows?

hafizrahman commented:
Thanks for the feedback @selanthiraiyan 

> I am worried about using break for actions with completion handlers. If we don't implement those actions by calling the completion handler, it will be difficult to debug it in future.

I followed the practice of adding cases into `break` from the  previous time the screenshot test was updated: https://github.com/woocommerce/woocommerce-ios/pull/6699/commits/4c0b1a3a6b88d445a786e6e357477c948e7dc3de

@rachelmcr outlined the rationale in that PR:

> Adds support for new actions in `MockAppSettingsActionHandler` and `MockStoresManager`. (These actions need to be handled, but specific behavior for these actions isn't required for the screenshots, so their cases are simply handled with a break.)

I think it's a delicate balance between getting the mock as accurate as possible, and getting the screenshot working as soon as possible. Do you think it makes sense to create a separate issue for this and work on it separately?

hafizrahman committed:
Remove unused property.

selanthiraiyan reviewed (approved):
> I think it's a delicate balance between getting the mock as accurate as possible, and getting the screenshot working as soon as possible. Do you think it makes sense to create a separate issue for this and work on it separately?

This sounds good to me. Thank you!

hafizrahman committed:
Update mock action handlers to avoid breaks.

hafizrahman commented:
@selanthiraiyan I ended up looking into it and it's not as time-consuming as I thought. Mocks are updated to avoid breaks in https://github.com/woocommerce/woocommerce-ios/pull/14335/commits/9e53017013e930e5b1e53106c4edfe034c16a3a8 and in my test it's not affecting the screenshots results.
